src/v/cluster/archival/archiver_operations_impl.cc

```diff
@@ -877,28 +884,358 @@ class archiver_operations_impl : public archiver_operations_api {
+    ss::future<upload_result> upload_stream(
+      const cloud_storage_clients::bucket_name& bucket,
+      cloud_storage_clients::object_key key,
+      uint64_t content_length,
+      ss::input_stream<char> stream,
+      cloud_storage::upload_type type,
+      retry_chain_node& parent) override {
+        // Without retries we don't have to use lazy abort source
+        lazy_abort_source noop(
+          []() -> std::optional<ss::sstring> { return std::nullopt; });
+
+        // Segment upload
+        auto path = cloud_storage::remote_segment_path(key());
+
+        auto handle_exceptions_fn = [type](ss::future<upload_result> fut) {
+            if (fut.failed()) {
+                const char* upl_type = "Segment";
+                if (type == cloud_storage::upload_type::segment_index) {
+                    upl_type = "Segment index";
+                }
+                vlog(
+                  archival_log.error,
+                  "{} upload failure, error: {}",
+                  upl_type,
+                  fut.get_exception());
+                return upload_result::failed;
+            }
+            return fut.get();
+        };
+        static constexpr size_t num_retries = 1;
+
+        auto result = upload_result::success;
+        if (type == cloud_storage::upload_type::segment_index) {
+            // The index object is small so it's OK to materialize it
+            iobuf payload;
+            auto pst = make_iobuf_ref_output_stream(payload);
+            co_await ss::copy(stream, pst);
+            auto& p = _remote.get_probe();
+            result = co_await _remote
+                       .upload_object(cloud_storage::upload_request{
+                         .transfer_details = {
+                           .bucket = bucket,
+                           .key = key,
+                           .parent_rtc = parent,
+                           .success_cb = [&p]() { p.index_upload(); },
+                           .failure_cb = [&p]() { p.failed_index_upload(); }},
+                         .type = cloud_storage::upload_type::segment_index,
+                         .payload = std::move(payload)})
+                       .then_wrapped(handle_exceptions_fn);
+        } else {
+            auto s = std::make_unique<one_time_stream_provider>(
+              std::move(stream));
+
+            auto reset_fn = [str = std::move(s)]() mutable {
+                return ss::make_ready_future<std::unique_ptr<stream_provider>>(
+                  std::move(str));
+            };
+            result = co_await _remote
+                       .upload_segment(
+                         bucket,
+                         path,
+                         content_length,
+                         std::move(reset_fn),
+                         parent,
+                         noop,
+                         num_retries)
+                       .then_wrapped(handle_exceptions_fn);
+        }
+
+        co_return result;
+    }
```

Comment:
I'll try but it might not be trivial because in this case the simplicity was sacrificed for efficiency. The idea is to upload segment and index by scanning the segment only once.